packages/astro/src/core/cookies/cookies.ts

```diff
@@ -0,0 +1,236 @@
+	 * request. If you have set a cookie via Astro.cookies.set(key, value), the value will be taken
+	 * from that set call, overriding any values already part of the request.
+	 * @param key The cookie to get.
+	 * @returns An object containing the cookie value as well as convenience methods for converting its value.
+	 */
+	get(key: string): AstroCookie {
+		// Check for outgoing Set-Cookie values first
+		if(this.#outgoing !== null && this.#outgoing.has(key)) {
+			let [serializedValue,, isSetValue] = this.#outgoing.get(key)!;
+			if(isSetValue) {
+				return new AstroCookie(serializedValue);
+			} else {
+				return new AstroCookie(undefined);
+			}
+		}
+
+		const values = this.#ensureParsed();
+		const value = values[key];
+		return new AstroCookie(value);
+	}
+
+	/**
+	 * Astro.cookies.has(key) returns a boolean indicating whether this cookie is either
+	 * part of the initial request or set via Astro.cookies.set(key)
+	 * @param key The cookie to check for.
+	 * @returns 
+	 */
+	has(key: string): boolean {
+		if(this.#outgoing !== null && this.#outgoing.has(key)) {
+			let [,,isSetValue] = this.#outgoing.get(key)!;
+			return isSetValue;
+		}
+		const values = this.#ensureParsed();
+		return !!values[key];
+	}
+
+	/**
+	 * Astro.cookies.set(key, value) is used to set a cookie's value. If provided
+	 * an object it will be stringified via JSON.stringify(value). Additionally you
+	 * can provide options customizing how this cookie will be set, such as setting httpOnly
+	 * in order to prevent the cookie from being read in client-side JavaScript.
+	 * @param key The name of the cookie to set.
+	 * @param value A value, either a string or other primitive or an object.
+	 * @param options Options for the cookie, such as the path and security settings.
+	 */
+	set(key: string, value: string | Record<string, any>, options?: AstroCookieSetOptions): void {
+		let serializedValue: string;
+		if(typeof value === 'string') {
+			serializedValue = value;
+		} else {
+			// Support stringifying JSON objects for convenience. First check that this is
+			// a plain object and if it is, stringify. If not, allow support for toString() overrides.
+			let toStringValue = value.toString();
+			if(toStringValue === Object.prototype.toString.call(value)) {
+				serializedValue = JSON.stringify(value);
+			} else {
+				serializedValue = toStringValue;
+			}
+		}
+
+		let expires: Date | undefined = undefined;
+		if(options?.expires) {
+			let rawExpires = options.expires;
+			switch(typeof rawExpires) {
+				case 'string': {
+					let numberOfMs = ms(rawExpires);
+					if(numberOfMs === undefined) {
+						if(rawExpires.includes('month')) {
+							throw new Error(`Cannot convert months because there is no fixed duration. Use days instead.`);
+						} else {
+							throw new Error(`Unable to convert expires expression [${rawExpires}]`);
+						}
+					}
+					let now = Date.now();
+					expires = new Date(now + numberOfMs);
+					break;
+				}
```

Comment:
I probably should've commented this in the RFC, but this feels like the user should use `Max-Age` instead if it's a duration.

Since we're aiming at a higher level implementation, maybe this is fine. Should we also supporting strings for `maxAge` too?

Reply:
It's ok to have further issues in the PR, this is a proper part of the RFC process! I think you raise a good point. Perhaps we should defer this part until later, we can add this as a future feature once we talk about it more.

Reply:
This was removed in https://github.com/withastro/astro/pull/4876/commits/515690c7a71d0bb5810fb68284117aa99aa628a1